Library: MQTT-SN Client library Gateway Search Support

SpontaneousDuck opened the pull request:
This pull requests extends the MQTT-SN Client library to support the "Gateway Advertisement and Discovery" process defined in section 6.1 of the MQTT-SN specification. This includes breaking changes to the transport interface and the default included UDP interface implementation as support for UDP multicast messages is added as implemented by the Paho MQTT-SN Gateway ([link](https://github.com/eclipse/paho.mqtt-sn.embedded-c)).

This PR closes issue #78010.

Changes List:
- [x] Change transport interface to support both addressed and broadcast messages.
- [x] Update the UDP example project to support the new transport interface and Gateway search process
- [x] Add decode code for SEARCHGW
- [x] Function for manually adding a gateway address\
- [x] Function to initialize Search process
- [x] Add handling for ADVERTISE packets
  - [x] Add GW to list upon reception
  - [x] Remove GW if miss set number of ADVERTISE
- [x] Add handling for SEARCHGW packets
  - [x] Respond after delay with GWINFO
  - [x] Cancel response if other responds first
- [x] Add handling for GWINFO packets
- [x] Connect will now use the first GW in the internal list
- [x] Time delay for responding to and transmitting SEARCHGW messages
- [x] Internal list for Gateways
- [x] Unit test coverage for new code

rlubos commented on subsys/net/lib/mqtt_sn/mqtt_sn_transport_udp.c:
```diff
@@ -43,25 +44,65 @@ static int tp_udp_init(struct mqtt_sn_transport *transport)
 
 	LOG_DBG("Socket %d", udp->sock);
 
+	/* Enable broadcasting mode */
+	/*
+	 * int broadcastEnable = 1;
```
Please remove commented out code.

SpontaneousDuck replied:
I was hoping to get the opinion of someone who knows Linux's networking stack better whether this is required or not? I was able to test it and it worked without this but unsure if that is correct or not.

rlubos replied:
Well, we don't even support `SO_BROADCAST` socket option in Zephyr so I guess the answer is it's not needed (at least with Zephyr stack).

rlubos commented on subsys/net/lib/mqtt_sn/mqtt_sn_transport_udp.c:
```diff
@@ -95,19 +159,33 @@ static int tp_udp_msg_send(struct mqtt_sn_client *client, void *buf, size_t sz)
+		return rc;
+	}
+
+	for (size_t i = 0; i < *addrlen; i++) {
+		if (srcaddr->data[i] != udp->bcaddr.data[i]) {
+			return rc;
+		}
+	}
```
Just use `memcmp()` perhaps?

rlubos commented on subsys/net/lib/mqtt_sn/Kconfig:
```diff
@@ -27,6 +27,18 @@ config MQTT_SN_LIB_MAX_TOPIC_SIZE
 
+config MQTT_SN_LIB_MAX_GATEWAYS
+	int "Maximum number of gateways to store internally"
+	default 2
```
This should've been done for other configs, but at least for new ones it'd be good to specify `range` parameter, so that users don't set for example negative values here.

SpontaneousDuck replied:
Is there a way to define open-ended ranges in Kconfig? Or should I just make the maximum a large/unrealistic number?

rlubos replied:
No open ranges, but we usually do something like this:
```
    range 1 $(UINT8_MAX)
```
Just choose the upper range to be something reasonable for particular config.

SpontaneousDuck replied:
Added range to all Kconfig for the mqtt_sn library

rlubos commented on samples/net/mqtt_sn_publisher/compose/README.md:
```diff
@@ -0,0 +1,6 @@
+This folder contains an example MQTT-SN Gateway made for use with a native_sim board application utilizing MQTT-SN.
```
Same here, convenience configs could be introduced in a separate commit.

rlubos reviewed (commented):
Thanks for the contribution, some initial feedback from my side. Looks pretty neat overall.

SpontaneousDuck commented on samples/net/mqtt_sn_publisher/src/udp.c:
```diff
@@ -109,27 +120,46 @@ static int do_work(void)
 	__ASSERT(err == 0, "mqtt_sn_client_init() failed %d", err);
 
+	/*
+	 * LOG_INF("Adding predefined Gateway");
+	 * struct sockaddr_in gwaddr = {0};
+	 * int err;
+	 * LOG_DBG("Parsing Broadcast IP " CONFIG_NET_SAMPLE_MQTT_SN_GATEWAY_IP);
+	 * gwaddr.sin_family = AF_INET;
+	 * gwaddr.sin_port = htons(CONFIG_NET_SAMPLE_MQTT_SN_GATEWAY_PORT);
+	 * err = inet_pton(AF_INET, CONFIG_NET_SAMPLE_MQTT_SN_GATEWAY_IP, &gwaddr.sin_addr);
+	 * __ASSERT(err == 1, "inet_pton() failed %d", err);
+	 * struct mqtt_sn_data gwaddr_data = {.data = (struct sockaddr *)&bcaddr,
+	 *				      .size = sizeof(struct sockaddr)};
+	 * err = mqtt_sn_add_gw(&mqtt_client, 0x1f, gwaddr_data);
+	 * __ASSERT(err == 0, "mqtt_sn_add_gw() failed %d", err);
```
Good idea! I'll go ahead and add that.

SpontaneousDuck commented on subsys/net/lib/mqtt_sn/mqtt_sn.c:
```diff
@@ -55,9 +56,23 @@ struct mqtt_sn_topic {
-K_MEM_SLAB_DEFINE_STATIC(publishes, sizeof(struct mqtt_sn_publish),
-			 CONFIG_MQTT_SN_LIB_MAX_PUBLISH, 4);
+struct mqtt_sn_addr {
+	char addr[CONFIG_MQTT_SN_LIB_MAX_ADDR_SIZE];
+	size_t addr_len;
```
I was expecting the addresses to be agnostic to the library and it would only store the address and pass pointers back to the transport code the user creates. Does the byte order matter then?

rlubos replied:
Sure, but the address format for a specific transport should still be known for the application, I'm not sure if it's documented anywhere right now.

SpontaneousDuck replied:
I could add some documentation to that that the user will need to define this based on their implemented transport?

rlubos replied:
That would be nice.

SpontaneousDuck commented on include/zephyr/net/mqtt_sn.h:
```diff
@@ -317,6 +346,29 @@ int mqtt_sn_client_init(struct mqtt_sn_client *client, const struct mqtt_sn_data
+ *
+ * @return 0 or a negative error code (errno.h) indicating reason of failure.
+ */
+int mqtt_sn_add_gw(struct mqtt_sn_client *client, uint8_t gw_id, struct mqtt_sn_data gw_addr);
```
I think the only difference between `struct mqtt_sn_addr` and `struct mqtt_sn_addr` is whether the struct contains the data or just a pointer. I can look at combining these two or making `struct mqtt_sn_addr` public and just using that.

SpontaneousDuck replied:
As I internally allocate memory to store the address, I removed `struct mqtt_sn_addr` entirely and only use `struct mqtt_sn_data` now.

jukkar commented on tests/net/lib/mqtt_sn_client/src/mqtt_sn_client.c:
```diff
@@ -122,24 +148,175 @@ static void mqtt_sn_connect_no_will(struct mqtt_sn_client *client)
+	zassert_equal(evt_cb_data.called, 3, "NO event");
+	zassert_equal(evt_cb_data.last_evt.type, MQTT_SN_EVT_CONNECTED, "Wrong event");
+
+	k_sleep(K_SECONDS(10));
```
The various sleeps in the tests are problematic. These just increase the CI pass time. Would it be possible to have for example a semaphore that waits, but is then released when the condition passes?

SpontaneousDuck replied:
Done! I removed all sleeps and replaced with semaphores. I also reduced the configured wait time to reduce test idle time.

jukkar commented on subsys/net/lib/mqtt_sn/mqtt_sn_transport_udp.c:
```diff
@@ -43,25 +44,65 @@ static int tp_udp_init(struct mqtt_sn_transport *transport)
 {
 	struct mqtt_sn_transport_udp *udp = UDP_TRANSPORT(transport);
 	int err;
+	struct sockaddr_in addrm;
```
So no IPv6 support?

SpontaneousDuck replied:
Not currently in the provided implementation but that matches the current implementation that I am extending. This does not preclude the user from implementing their own transport that uses IPv6. Should I add support as part of this PR or leave the current implementation's limitation?

jukkar replied:
As the other parts of the Zephyr network stack typically support IPv6 and IPv4 (dual stack), it would be preferable if all new code also support dual stack. So yes, I would prefer IPv6 support also in this PR.

jukkar commented on subsys/net/lib/mqtt_sn/mqtt_sn_transport_udp.c:
```diff
@@ -43,25 +44,65 @@ static int tp_udp_init(struct mqtt_sn_transport *transport)
 #ifdef LOG_DBG
 	char ip[30], *out;
 
-	out = get_ip_str((struct sockaddr *)&udp->gwaddr, ip, sizeof(ip));
+	out = get_ip_str((struct sockaddr *)&udp->bcaddr, ip, sizeof(ip));
 	if (out != NULL) {
-		LOG_DBG("Connecting to IP %s:%u", out,
-			ntohs(((struct sockaddr_in *)&udp->gwaddr)->sin_port));
+		LOG_DBG("Binding to Brodcast IP %s:%u", out,
+			ntohs(((struct sockaddr_in *)&udp->bcaddr)->sin_port));
 	}
 #endif
```
My comment is not directly related to your changes. It would actually be better to use the actual debug enabling option here.
So something like
```
if (IS_ENABLED(CONFIG_MQTT_SN_LOG_LEVEL_DBG)) {
....
}
```
This way the debugging code will be compile tested but will not be enabled unless the debugging option is enabled.

SpontaneousDuck commented on subsys/net/lib/mqtt_sn/Kconfig:
```diff
@@ -27,6 +27,18 @@ config MQTT_SN_LIB_MAX_TOPIC_SIZE
+
+config MQTT_SN_LIB_MAX_ADDR_SIZE
+	int "Maximum address size for the transport"
+	default 21
```
The problem is that some user could implement custom transports (as I am planning on doing) which does not use the standard addressing schemes. I was aiming for maximum flexibility for users going forward to not limit them to the IPv4 stack that is provided more as an example.

rlubos replied:
Ok, fine then.

SpontaneousDuck commented:
What would be the recommended way to add changes based on code comments? It appears like the desired commits to merge will be all the code changes in one commit with test cases and convenience commits in different commits (3 total). Is this incorrect and I can add the requested changes in another commit? Or should I `git reset` and `git push --force` to get the desired number of commits which will break the comments left on this PR?

rlubos commented:
> What would be the recommended way to add changes based on code comments? It appears like the desired commits to merge will be all the code changes in one commit with test cases and convenience commits in different commits (3 total). Is this incorrect and I can add the requested changes in another commit? Or should I `git reset` and `git push --force` to get the desired number of commits which will break the comments left on this PR? I have a bunch of changes due to above comments but am not sure what the recommended way to push them to the tree is.

Typically you just squash the review fixes into original commits (you can use `git commit --fixup <original commit SHA>` and then interactive reabase with `--auto-squash` option).

As for splitting the changes into separate commits, I guess the easiest would be to `git reset` and then create separate commits.

And yes, you'd need to force-push as the history is rewritten, but that's a common thing to do in Zephyr PRs. Don't worry about review comments, GH will handle this.

SpontaneousDuck commented:
Ok, I think I made it through all the comments and pushed all the changes up in 3 new commits. One for main code and doc changes, one with unit tests, and one with the sample code updates.

jukkar commented on subsys/net/lib/mqtt_sn/mqtt_sn_transport_udp.c:
```diff
@@ -43,25 +44,83 @@ static int tp_udp_init(struct mqtt_sn_transport *transport)
+
+	switch (udp->bcaddr.sa_family) {
+	case AF_INET:
+		addrm.sa_family = AF_INET;
+		((struct sockaddr_in *)&addrm)->sin_port =
+			((struct sockaddr_in *)&udp->bcaddr)->sin_port;
+		((struct sockaddr_in *)&addrm)->sin_addr.s_addr = INADDR_ANY;
```
Is IPv4 always enabled, if it is, then we should have `depends on NET_IPV4` in Kconfig file.
Otherwise we should place `IS_ENABLED(CONFIG_NET_IPV4)` here like in IPv6.

jukkar commented on subsys/net/lib/mqtt_sn/mqtt_sn_transport_udp.c:
```diff
@@ -43,25 +44,83 @@ static int tp_udp_init(struct mqtt_sn_transport *transport)
-		LOG_DBG("Connecting to IP %s:%u", out,
-			ntohs(((struct sockaddr_in *)&udp->gwaddr)->sin_port));
+	memcpy(&mreqn.imr_multiaddr, &udp->bcaddr.data[2], sizeof(udp->bcaddr.data) - 2);
+	iface = net_if_get_default();
```
This might give wrong interface if there are multiple in the system. Perhaps we could select the one where the gwaddr traffic is routed. You can use `net_if_ipv6_select_src_iface()` for IPv6 and `net_if_ipv4_select_src_iface()` for IPv4 to find out the correct interface.

jukkar commented on samples/net/mqtt_sn_publisher/Kconfig:
```diff
@@ -5,12 +5,21 @@
+
 config NET_SAMPLE_MQTT_SN_GATEWAY_IP
-	string "IP of the MQTT-SN gateway"
+	string "IP of the MQTT-SN gateway. Only used if NET_SAMPLE_MQTT_SN_STATIC_GATEWAY=n."
```
You could put `depends on` here for the `NET_SAMPLE_MQTT_SN_STATIC_GATEWAY` to hide the symbol so user would not be able to set the value.

SpontaneousDuck replied:
So I did try that. In samples/net/mqtt_sn_publisher/src/udp.c:141 I use `if (IS_ENABLED(CONFIG_NET_SAMPLE_MQTT_SN_STATIC_GATEWAY)) {` with `CONFIG_NET_SAMPLE_MQTT_SN_GATEWAY_IP` referenced as requested earlier. If I make `NET_SAMPLE_MQTT_SN_GATEWAY_IP` depend on `NET_SAMPLE_MQTT_SN_STATIC_GATEWAY` then when `NET_SAMPLE_MQTT_SN_STATIC_GATEWAY=n`, `NET_SAMPLE_MQTT_SN_GATEWAY_IP` is undefined and build fails. Should I put a `#ifdef` around that part as well or is there a `IS_ENABLED` analog for if a CONFIG is defined at all?

jukkar replied:
You could do for example like this
```
#ifdef CONFIG_NET_SAMPLE_MQTT_SN_STATIC_GATEWAY
#define SAMPLE_GW_IP CONFIG_NET_SAMPLE_MQTT_SN_STATIC_GATEWAY_IP
#else
#define SAMPLE_GW_IP ""
#endif 
```
and then use `SAMPLE_GW_IP` when referring to the config value.
Alternatively there is also `IF_EXISTS()` macro that could be used here.

SpontaneousDuck committed:
net: mqtt-sn: Add Gateway Advertisement and Discovery process support

Fixes: #78010
This commit implements the "Gateway Advertisement and Discovery" process
defined in section 6.1 of the MQTT-SN specification.
This includes breaking changes to the transport interface and the default
included UDP interface implementation as support for UDP multicast
messages is added as implemented by the Paho MQTT-SN Gateway.

Signed-off-by: Kenneth Witham <kennywitham4@gmail.com>

SpontaneousDuck committed:
net: mqtt-sn: Update MQTT-SN Unit Tests

Fixes: #78010
This commit adds tests for the "Gateway Advertisement and Discovery"
process. This includes tests for handling SEARCHGW, GWINFO, and ADVERTISE
as well as manually adding a Gateway for use. Gateway pruning when
missing ADVERTISE messages is also tested.

Signed-off-by: Kenneth Witham <kennywitham4@gmail.com>

SpontaneousDuck committed:
net: mqtt-sn: Update MQTT-SN Publisher Sample with Gateway Discovery

Fixes: #78010
This commit adds the "Gateway Advertisement and Discovery" process
to the MQTT-SN Publisher Sample application.

Signed-off-by: Kenneth Witham <kennywitham4@gmail.com>

rlubos reviewed (dismissed):
Just some nits spotted, but other wise no more comments from my side, looks good.

jukkar reviewed (dismissed):
LGTM

SpontaneousDuck commented:
Just quickly fixed the nit posted above

kartben commented on doc/connectivity/networking/api/mqtt_sn.rst:
```diff
@@ -84,15 +84,31 @@ used. An example configuration for UDP transport is shown below:
+discover gateways through an advertisement or a search mechanism. A user
+should do at least one of the following steps to define a Gateway for the library:
+
+* Call the ``mqtt_sn_add_gw`` function to manually define a Gateway address.
+* Wait for a ``MQTT_SN_EVT_ADVERTISE``.
+* Call the ``mqtt_sn_search`` function and wait for a ``MQTT_SN_EVT_GWINFO`` callback.
+  Make sure to call the ``mqtt_sn_input`` function periodically to process incoming messages.
+
+Example ``mqtt_sn_search`` function call:
+
+.. code-block:: c
+
+	err = mqtt_sn_search(&mqtt_client, 1);
+	k_sleep(K_SECONDS(10));
+	err = mqtt_sn_input(&mqtt_client);
+	__ASSERT(err == 0, "mqtt_sn_search() failed %d", err);
+
+After the Gateway address has been defined or found, the MQTT-SN client can
+connect to the gateway. Call the ``mqtt_sn_connect`` function, which will send a
+``CONNECT`` message. The application should periodically call the ``mqtt_sn_input``
+function to process the response received. The application does not have to call
+``mqtt_sn_input`` if it knows that no data has been received (e.g. when using Bluetooth).
+Note that ``mqtt_sn_input`` is a non-blocking function, if the transport struct contains a
```
```suggestion
* Call the :c:func:`mqtt_sn_add_gw` function to manually define a Gateway address.
* Wait for an :c:enumerator:`MQTT_SN_EVT_ADVERTISE`.
* Call the :c:func:`mqtt_sn_search` function and wait for an :c:enumerator:`MQTT_SN_EVT_GWINFO` callback.
  Make sure to call the :c:func:`mqtt_sn_input` function periodically to process incoming messages.

Example :c:func:`mqtt_sn_search` function call:

.. code-block:: c

	err = mqtt_sn_search(&mqtt_client, 1);
	k_sleep(K_SECONDS(10));
	err = mqtt_sn_input(&mqtt_client);
	__ASSERT(err == 0, "mqtt_sn_search() failed %d", err);

After the Gateway address has been defined or found, the MQTT-SN client can
connect to the gateway. Call the :c:func:`mqtt_sn_connect` function, which will send a
``CONNECT`` message. The application should periodically call the :c:func:`mqtt_sn_input`
function to process the response received. The application does not have to call
:c:func:`mqtt_sn_input` if it knows that no data has been received (e.g. when using Bluetooth).
Note that :c:func:`mqtt_sn_input` is a non-blocking function, if the transport struct contains a
```

BTW I appreciate that the rest of the file probably has other instances of C functions/enumerators incorreclty referenced, but at least it would be good to get these ones right (and maybe you'll feel like making a full pass on the file as a follow-up :-))

SpontaneousDuck replied:
Thanks for the feedback! I have never worked with RST files before so I am learning along the way. Made a full pass on the file and hope it is correct now! :)

kartben commented on samples/net/mqtt_sn_publisher/README.rst:
```diff
@@ -40,13 +40,29 @@ Requirements
-build it from source - see `PAHO MQTT-SN Gateway`_ - or run an unofficial
-docker image, like `kyberpunk/paho`_.
+You will also need to start an MQTT-SN gateway. A convenience Docker Compose specification file
+is provided in :file:`compose/docker-compose.yaml`. First, Start the net-tools configuration from
```
typo in file name (you may actually want to call it docker-compose.yaml instead of just compose.yaml?), + use zephyr_file for it to be a proper link (and, incidentally, check that the referenced file actually exists :))

```suggestion
is provided in :zephyr_file:`samples/net/mqtt_sn_publisher/compose/compose.yaml`. First, Start the net-tools configuration from
```

SpontaneousDuck replied:
Added the `zephyr_file` change. Looked it up and `compose.yaml` is the recommended file name by Docker so leaving it at that :)

kartben reviewed (changes requested):
some comments on the contributed documentation. Thanks for the PR!

kartben commented:
@SpontaneousDuck we are seeing errors our weekly CI run which I suspect could be due to this PR. Would you be able to have a look by any chance?
https://github.com/zephyrproject-rtos/zephyr/actions/runs/11992688445/job/33432830180

SpontaneousDuck commented:
Will do! Looking now

SpontaneousDuck commented:
Ok, I found the issue. I am calling `LOG_ERR` with a `%" PRIu16 "` when the printed data is a `long unsigned int`. The default platforms tested (`native_sim`, `qemu_cortex_a9`, `qemu_x86`, `mps2/an385`) do not throw errors for this but `bcm958402m2/bcm58402/a72` does. 

There are no proper types in `inttypes.h` that results in the `%lu` format specification. I just need to update the `LOG_ERROR` calls to `%lu` and the tests pass. 

How should I make the update? Should I make a new Issue and PR or just fixup the commit and reopen this PR somehow?

kartben commented:
> Ok, I found the issue. I am calling `LOG_ERR` with a `%" PRIu16 "` when the printed data is a `long unsigned int`. The default platforms tested (`native_sim`, `qemu_cortex_a9`, `qemu_x86`, `mps2/an385`) do not throw errors for this but `bcm958402m2/bcm58402/a72` does. 
> 
> There are no proper types in `inttypes.h` that results in the `%zu` format specification. I just need to update the `LOG_ERROR` calls to use `%zu` and the tests pass. 
> 
> How should I make the update? Should I make a new Issue and PR or just fixup the commit and reopen this PR somehow? @kartben 

The latter, PR has been merged already so it's too late for a fixup :) thanks for the quick turnaround!

SpontaneousDuck commented:
Fixed in #82007
